### DISCUSSION TITLE
Already ready for an "alpha" implementation on a texture website?

### DISCUSSION THREAD

**Discussion by struffel** [OpenAssetIO/OpenAssetIO#831] [General]
Hello everyone, I found out about OpenAssetIO a while ago and it actually looks really interesting. People have been asking me about writing Plug-ins for software like Blender, Unity, Unreal Engine, Maya and others to make importing materials from my website ambientcg.com more comfortable and I have always been reluctant about it because it's a lot of work to continuously support and update these plugins.

Now if I understand OpenAssetIO correctly then it has the potential to make these plugins redundant because I "only" need to put an "OpenAssetIO Host" (if that's the proper name) publicly on the internet and then people can connect with my service and fetch assets from my website (it's all open and Public Domain so there is no auth required). But looking at the current documentation and examples I am not sure how that would work in detail or if I have even undrestood the principle correctly.

Is it already possible for someone to create the kind an "OpenAssetIO Host" on the public internet, and if so, how would that work in detail? Since ambientCG.com is run by only me it is very flexible and I would be happy to be an "early adopter", even if that means that the system doesn't work as well in the beginning. But I don't quite know where to start.

**Comment by foundrytom** [comment]
Hello @struffel!

This is a great question. It would be really good to look into what supporting Ambient CG would be like. We've always wanted the project to simplify the integration of services like yours as well as the more 'traditional' management systems like ShotGrid et. al. If you're able to spend some time with us on this, its a perfect time. We're aiming for some form of a 'v1' later in the year - so discovering any specific issues when working with this kind of library before then makes it much easier to address them in the API itself if needed.

> Now if I understand OpenAssetIO correctly then it has the potential to make these plugins redundant because I "only" need to put an "OpenAssetIO Host"

This is our (naive) goal - and we live in the hope we can help make this it nothing else, more possible than it is today.

> "OpenAssetIO Host" (if that's the proper name) publicly on the internet and then people can connect with my service and fetch assets from my website

Yeah - though its the other way around - you'd be fulfilling the Manager part of the equation, (as you "manage" the materials, etc).

> Is it already possible for someone to create the kind an "OpenAssetIO Host" on the public internet, and if so, how would that work in detail?

Because of the nature of your service, there are a few more moving parts that we'd need to consider. It'll attempt to sketch out how it could/would work here:

The API is deliberately designed to provide 'practical working code' over just a specification. This means that (right now) you need to provide a `ManagerPlugin` implementation that is capable of fulfilling the API through your back-end. For the software you mentioned, this would need to be written Python or C++. This would then talk to your site however you wished.

Longer term, we have been looking at using [gRPC](https://github.com/foundrytom/OpenAssetIO-gRPC) or [OpenAPI](https://github.com/foundrytom/OpenAssetIO-OpenAPI) to avoid the need for a custom plugin, but this is along long way off being production ready and needs a lot of ⌨️ .

Having a plugin would enable several things (long term):
  - Users to reference/query things from your library via ID, vs file path [doable now]
  - Publish new things to your library (if desired) [doable now]
  - Users to use your library browser directly in an application [The UI layer is Q4Y23+]

There is an extra challenge for you here though. Our first set of target systems currently rely on the data that any given plugin is managing being "available to the host in some suitable form". The simple example of this is that the library resolves to a file system path that the application/DCC tool has access to - so loading the actual bytes is effectively the same as the "bad old days".

This isn't the case when the data is held on a web site as it is in your case. And so right now, unless the DCC itself supports direct access via HTTP, your plugin would have to take care of downloading the data for the user, managing that on disk, and resolving the URI to wherever that is when the application needs it. With the opposite happening in reverse during publish.

Making this better, is something that is independent of OpenAssetIO - and we're trying to take some steps to improve DCCs [with  this proposal](https://github.com/foundrytom/proposals/pull/1), but I havent had any time to push that forward lately. Having any solid, unilateral support for non file based resources in DCCs feels like its a while off yet so that download work falls back into our collective 'pipeline' camp for a while yet I think unfortunately. A community 'Download manager' project would be amazing, if it could be easily integrated into peoples OpenAssetIO plugins where they deal with alternate storage infrastructure.

Once we have a working plugin, there is potentially another step to consider - in that it would most likely be used in conjunction with other management systems (as its not an all encompassing production AMS). For a long time we've planned a 'fan out' manager adapter to make multiple managers relatively straight forward, so the DCCs don't need to worry too much. We haven't done this yet though, and I don't think it happen this year without any extra typing. Nothing stopping people using it without, just it would be the only manager available to them.

So in summary. Theoretically, its all possible and fits within the design. There is however a lot of work to be  done. On top of everything else I mentioned, we then need the tools you listed to support the API too! We're in a bit of a 🐔 🥚 scenario right now, as without the Managers ,the Hosts are reluctant, and without the Hosts, the Managers are reluctant 🤣. We're hoping to break that stalemate a bit now as we're starting work on Nuke and Katana integration at Foundry.

We'd really love to work through this all though as ultimately, integrating services like yours is exactly what we want to be able to support in the long term...

Hope that made some sense, and let us know what you think!

Thanks again for reaching out. Always happy to jump on a call or similar if it saves the typing.

**Reply by struffel** [reply]
Thank you for your detailed response and I'm sorry for never responding to this but I only dove back into this topic over the holidays.
This all sounds very interesting, though also challenging. Especially that OpenAPI snippet looks extremely interesting, being able to provide a standard JSON API to turn my server into a "compliant" OpenAssetIO Asset Manager and leaving the rest to the community of the individual DCCs would be a dream.

Recently, I have again experimented with designing a general-purpose "asset-fetching" API, that could basically make "offerings" of multiple implementations (like .blend files for blender, .max files for 3DSMax and .obj for everyone else etc.) from which the client (i.e. the plugin inside the DCC) can then pick one that it can actually load. Looking at what's currently there I wonder if it is possible to create some kind of "request form", that allows users to specify something like "give me all the materials tagged 'paving' with 2K resolution"?

Another aspect is how the assets get treated after the initial hand-over. You already talked about how I might need to download the assets myself using python which I wouldn't consider a huge hurdle. But I wonder how that transition from a remote asset to a local one would work. If an artist in a studio decides to integrate `ambientCG:///PavingStones036/PavingStones036_2K-PNG` (`locatableContent` points to USDZ file with one material and the actual maps) into their project, is there a way to turn that into a new naming internal entity, for example `localManager:///Projects/XY/Materials/PavingStonesMossy` so that the DCC knows to no longer ask the ambientCG Server about this file, because it has nothing to do with it anymore? I have seen the wonderfully named [Deassetization](https://learn.foundry.com/nuke/developers/latest/pythondevguide/openassetio.html#deassetization) in Nuke, but this would require something like a "Reassetization".

Also, just to make sure I understand this, I can't really do my own OpenAssetIO plugin for Blender (most requested by my community) until the Blender Developers have built the OpenAssetIO "foundation" into the tool itself, right?
Because if that is the case then I might release my asset-fetching API as a standalone JSON API for which anyone can built a client plugin or server implementation for but make sure to structure the data in such a way that it can easily be "ported" to run on top of OpenAssetIO.

**Reply by feltech** [reply]
Hello Lennart,

Tom may reply, but is no longer involved in day-to-day development on OpenAssetIO, so I'll take a stab.

> Looking at what's currently there I wonder if it is possible to create some kind of "request form", that allows users to specify something like "give me all the materials tagged 'paving' with 2K resolution"?

To me this sounds like a relationship query. We have methods that hosts (DCCs) can call (e.g. `getWtihRelationship`), where, given a "root" entity reference, you can query "related" references.  The relationship is described using traits, which are essentially tags with properties.  We would need well-defined traits for defining the metadata you want to filter by.  We have the public community-sourced [MediaCreation traits](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/blob/main/traits.yml) (see those with a `usage` of `relationship`), which can be extended. It's also possible to define your own traits, though the host application must know about them, somehow.

> is there a way to turn that into a new naming internal entity, for example `localManager:///Projects/XY/Materials/PavingStonesMossy` so that the DCC knows to no longer ask the ambientCG Server about this file

This particular example seems to imply that there is a second asset management system (`localManager`), which has intercepted/proxied the `ambientCG` entity resolve/download and cached on-site.  I can't think of a mechanism we have right now to change the entity reference as part of a `resolve` call.  Though, if you had a proxy manager, then it could cache the fact that the download has already taken place and there would be no need to change the entity reference.

Thinking more local to the workstation, an ambientCG manager plugin could implement some caching, knowing not to bother the remote server if the file had already been downloaded.  The DCC tool always just `resolve`s a reference to a file on disk, regardless what the plugin does in the background.

Using "pure" OpenAssetIO, the host application could `resolve` from ambientCG, then publish to localManager.  In the publishing workflow you  `register` a reference, which returns a "final" reference.  But the host application would have to know to do that workflow. 

> Also, just to make sure I understand this, I can't really do my own OpenAssetIO plugin for Blender (most requested by my community) until the Blender Developers have built the OpenAssetIO "foundation" into the tool itself, right?

Yeah that's right. They have to integrate the OpenAssetIO core library as a "host" application.
